### PR TITLE
Implement DNS-01 challenge

### PR DESCRIPTION
**Pull request by syberalexis**
Add implementation for DNS-01 challenge and enable it
Add pytests for the new implementation

Refer to #8

### CODE REVIEW

File: serles/challenge.py [hunk 1]
```diff
@@ -2,9 +2,11 @@
 import json
 import socket
 import hashlib
+import base64
 import requests
 import jwcrypto.jwk  # fedora package: python3-jwcrypto.noarch
 import jwcrypto.jws
+import dns.resolver
```
Comment (uedvt359) [inline serles/challenge.py:9]: new requirement needs to be added to setup.py

File: serles/challenge.py [hunk 2]
```diff
@@ -127,6 +131,46 @@ def http_challenge(challenge):  # RFC8555 §8.3
...
+    # Verify the expected challenge is present
+    expect = calculate_sha256(challenge)
+    is_expected = False
+    for answer in answers:
+        if expect == str(answer).strip('"'):
+            is_expected = True
+            break
+
+    if not is_expected:
+        return "incorrectResponse", f"no token found in TXT record {expect}"
```
Comment (uedvt359) [inline serles/challenge.py:171]:
```suggestion
    for answer in answers:
        if expect == answer.strings[0].decode():
            break
    else:
        return "incorrectResponse", f"no token found in TXT record {expect}"
```
(yes, that's the correct indentation for this else :) )

the `.strip('"')` makes me a bit queasy. i believe we can get to the challenge string with the code above, although we need to make sure it doesn't raise an exception (if .strings is empty or contains non-ascii characters)

File: serles/challenge.py [hunk 3]
```diff
@@ -127,6 +131,46 @@ def http_challenge(challenge):  # RFC8555 §8.3
...
+    """
+    host = challenge.authorization.identifier.value
+
+    # When it´s a wildcard domain, remove the wildcard
+    if host.startswith("*."):
+        host = host[2:]
+
+    # Try to resolve the _acme-challenge record
+    try:
+        answers = dns.resolver.query(f"_acme-challenge.{host}", "TXT")
```
Comment (uedvt359) [inline serles/challenge.py:152]:
```
>>> dns.resolver.query("tirol.gv.at", "TXT")
<stdin>:1: DeprecationWarning: please use dns.resolver.resolve() instead
```

File: serles/challenge.py [hunk 4]
```diff
@@ -188,7 +232,6 @@ def alpn_challenge(challenge):  # RFC 8737 §3
 
     return None, None  # no error occurred :)
 
-
```
Comment (uedvt359) [inline serles/challenge.py:191]: nit-pick: please leave functions double-spaced. also applies to new functions added here.

File: serles/challenge.py [hunk 5]
```diff
@@ -263,6 +305,18 @@ def key_authorization(challenge):
     ).thumbprint()
     return f"{token}.{thumbprint}"
 
+def calculate_sha256(challenge):
```
Comment (uedvt359) [inline serles/challenge.py:308]: i'd prefer to inline this function, since it is specific to the dns challenge

File: tests/test_challengefuns.py [hunk 6]
```diff
@@ -210,16 +239,17 @@ def test_verify_challenge_invalid(self):
...
-        with unittest.mock.patch.object(
-            mock_challenge, "type", main.ChallengeTypes.dns_01
-        ):
-            self.assertRaisesRegex(
-                main.ACMEError,
-                "challenge type not supported",
-                main.verify_challenge,
-                mock_challenge,
-            )
+    # No challenge are unsupported now !
```
Comment (uedvt359) [inline tests/test_challengefuns.py:242]: don't think we need this any more, please just remove it.

File: serles/challenge.py [hunk 7]
```diff
@@ -127,6 +131,46 @@ def http_challenge(challenge):  # RFC8555 §8.3
...
+    Args:
+        challenge (Challenge): The DNS-01 challenge to verify.
+
+    Returns:
+        tuple(str,str): problem detail type of the error and  textual
+        description, or (None,None).
+    """
+    host = challenge.authorization.identifier.value
+
+    # When it´s a wildcard domain, remove the wildcard
```
Comment (uedvt359) [inline serles/challenge.py:146]: i believe we need special handling for wildcard certs elsewhere, too. i'd consider this out of scope of this pr, though; we can handle it later (it should at least be a separate commit, if you intend to include it in this PR).

ctrl-f for "wildcard" yields results in:
https://datatracker.ietf.org/doc/html/rfc8555#section-7.1.3
https://datatracker.ietf.org/doc/html/rfc8555#section-7.1.4

**Review by uedvt359 — CHANGES_REQUESTED** [review]
Thank you a lot! I only had time for a cursory look, but it looks well-implemented. I appreciate you implementing tests as well. I'll need to run the test suite and compare it against the specification, still.

i've added some inline notes. for further changes, i'd prefer rebased commits, if you can.

File: serles/challenge.py (hunk 1 above)
Comment (syberalexis) [reply serles/challenge.py:9]: You already add it with dnspython package

Comment (uedvt359) [reply serles/challenge.py:9]: oops, completely forgot we already depend on that. sorry for the noise

File: serles/challenge.py (hunk 2 above)
Comment (syberalexis) [reply serles/challenge.py:171]: Your `else` proposal don't work if DNS return multiple TXT !? It's return error at the first answer if it don´t match.
`makes me a bit queasy`... I'm not a Python developer...
I didn't know string `decode` function, thanks.
I don't understand how works `.strings[X]`, i trust you. But managing strings with array index like this is a very bad practice, possibly a lib or language failure, idk...

File: serles/challenge.py (hunk 3 above)
Comment (syberalexis) [reply serles/challenge.py:152]: Oh i didn't see this, tkanks

File: serles/challenge.py (hunk 4 above)
Comment (syberalexis) [reply serles/challenge.py:191]: Ok, if you want ^^

File: serles/challenge.py (hunk 2 above)
Comment (uedvt359) [reply serles/challenge.py:171]: > Your else proposal don't work if DNS return multiple TXT

yes, it does. this is a for-else loop, not an if-else statement. it's a lesser used feature, but in short: if the loop is not `break`ed, the else clause of the for-loop is triggered.

`.strings` is an [attribute of the dnspython response object](https://dnspython.readthedocs.io/en/stable/rdata-subclasses.html#dns.rdtypes.ANY.TXT.TXT.strings), which is a list of the content.

Comment (syberalexis) [reply serles/challenge.py]: Obviously, `answers` is returned as an array of string with `resolve()` instead of `query()`

**Review by uedvt359 — CHANGES_REQUESTED** [review]
thanks for the quick update. we're getting really close. i still haven't run your code, but i only have one remaining piece of feedback before i'll start testing it out.

File: serles/challenge.py [hunk 8]
```diff
@@ -128,6 +132,42 @@ def http_challenge(challenge):  # RFC8555 §8.3
...
+    except dns.resolver.YXDOMAIN as e:
+        return "dnsQueryTooLong", e
+    except dns.resolver.LifetimeTimeout as e:
+        return "dnsTimeout", e
+
+    # Verify the expected challenge is present
+    sha256_digest = hashlib.sha256(key_authorization(challenge).encode('utf-8')).digest()
+    expect = base64.urlsafe_b64encode(sha256_digest).decode('utf-8').rstrip("=")
+    for answer in answers:
+        if expect == answer:
```
[the remark this answers is not in the document]
Comment (syberalexis) [reply serles/challenge.py:163]: Hum ok :)

I think, it's my data tests that fuck my brain with doesn't use `.strings`
Oh yes, i see it, i don't make the link with `.string[0]`, it's good for me :)

**Comment by uedvt359** [comment]
i have installed your branch on one of our development machines and successfully generated a certificate using it! this makes me very happy :)

```
[root@someserver ~]# REQUESTS_CA_BUNDLE=/etc/pki/tls/cert.pem /usr/bin/certbot certonly --config /etc/letsencrypt/cli-rzapi-someserver.ini --force-renewal  --manual --preferred-challenges dns -v
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator manual, Installer None
Renewing an existing certificate for someserver.example.com
Performing the following challenges:
dns-01 challenge for someserver.example.com

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Please deploy a DNS TXT record under the name:

_acme-challenge.someserver.example.com.

with the following value:

nfTaYBNG7Dj6BB470-_o808wAMRvzIa3ssE1gR1xx48

Before continuing, verify the TXT record has been deployed. Depending on the DNS
provider, this may take some time, from a few seconds to multiple minutes. You can
check if it has finished deploying with aid of online tools, such as the Google
Admin Toolbox: https://toolbox.googleapps.com/apps/dig/#TXT/_acme-challenge.someserver.example.com.
Look for one or more bolded line(s) below the line ';ANSWER'. It should show the
value(s) you've just added.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Press Enter to Continue
Waiting for verification...
Cleaning up challenges

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/someserver.example.com/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/someserver.example.com/privkey.pem
This certificate expires on 2025-02-03.
These files will be updated when the certificate renews.

NEXT STEPS:
- This certificate will not be renewed automatically. Autorenewal of --manual certificates requires the use of an authentication hook script (--manual-auth-hook) but one was not provided. To renew this certificate, repeat this same certbot command before the certificate's expiry date.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[root@someserver ~]# cat /etc/letsencrypt/cli-rzapi-someserver.ini
server = https://serles.example.com:8443/directory
domains = someserver.example.com
register-unsafely-without-email = True
keep-until-expiring = True
agree-tos = True
key-type = rsa
rsa-key-size = 4096
```

File: tests/test_challengefuns.py [hunk 9]
```diff
@@ -284,6 +321,98 @@ def test_http_challenge_ptr(self):
...
+        mock_dns_challenge.authorization.identifier.value = "NoNameservers"
+        with unittest.mock.patch.object(
+            dns.resolver, "query", MockedDNSResolveTXT
+        ), unittest.mock.patch.object(
+            dns.resolver, "resolve", MockedDNSResolveTXT
+        ):
+            result = main.dns_challenge(mock_dns_challenge)
+            self.assertEqual(result[0], "dnsNoNameServers")
+
+    def test_dns_challenge_nonameservers(self):
```
[the remark this answers is not in the document]
Comment (syberalexis) [reply tests/test_challengefuns.py:395]: Arf, i don't see that

File: tests/test_challengefuns.py [hunk 10]
```diff
@@ -210,17 +258,6 @@ def test_verify_challenge_invalid(self):
             )
             mock_challenge.authorization.order.authorizations = old_authz
```
[the remark this answers is not in the document]
Comment (syberalexis) [reply tests/test_challengefuns.py:260]: Ah yes, i don´t see that !
Thanks for the html generating command

**Comment by uedvt359** [comment]
Thanks, great work!

**Comment by syberalexis** [comment]
Thank you for corrections and help :)

**Pull request merged**